python/ray/data/_internal/execution/operators/map_operator_state.py

```diff
@@ -201,3 +218,62 @@ class _TaskState:
 
     inputs: RefBundle
     output: Optional[RefBundle] = None
+
+
+class _OutputQueue:
+    """Interface for swapping between different output order modes."""
+
+    def add_pending_task(self, task: "_TaskState"):
+        """Track the addition of a new task."""
+        pass
+
+    def notify_task_completed(self, task: "_TaskState"):
+        """Called when a previously added task completes."""
+        pass
+
+    def has_next(self) -> bool:
+        raise NotImplementedError
+
+    def get_next(self) -> RefBundle:
+        raise NotImplementedError
+
+
+class _OrderedOutputQueue(_OutputQueue):
+    """An queue that returns finished tasks in submission order."""
+
+    def __init__(self):
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._next_task_index: int = 0
+        self._next_output_index: int = 0
+
+    def add_pending_task(self, task: "_TaskState"):
+        self._tasks_by_output_order[self._next_task_index] = task
+        self._next_task_index += 1
+
+    def has_next(self) -> bool:
+        i = self._next_output_index
+        return (
+            i in self._tasks_by_output_order
+            and self._tasks_by_output_order[i].output is not None
+        )
+
+    def get_next(self) -> RefBundle:
+        i = self._next_output_index
+        self._next_output_index += 1
+        return self._tasks_by_output_order.pop(i).output
+
+
+class _UnorderedOutputQueue(_OutputQueue):
+    """An queue that does not guarantee output order of finished tasks."""
+
+    def __init__(self):
+        self._completed_tasks: List[_TaskState] = []
```

Comment:
This queue also preserves the order, just for notification order. Do we need these two subclasses?

Reply:
The order that matters here is pending order. Notification order is nondeterministic ("unordered").